File: src/sync/merge.ts
/**
 * Conflict detection for sync operations.
 */

import type { SyncMeta, SyncStatusEntry, SyncConflict } from "./types.js";

/**
 * Map dataKey to the corresponding version field in SyncMeta.
 */
function getLocalVersion(
  meta: SyncMeta | null,
  dataKey: string
): number {
  if (!meta) return 0;

  if (dataKey === "config") return meta.configVersion;
  if (dataKey === "providers") return meta.providersVersion;
  return 0;
}

/**
 * Detect conflict between local and remote versions for a given data key.
 *
 * Returns null if no conflict (safe to push/pull).
 * Returns SyncConflict if versions diverge.
 */
export function detectConflict(
  localMeta: SyncMeta | null,
  remoteStatuses: SyncStatusEntry[],
  dataKey: string
): SyncConflict | null {
  const localVersion = getLocalVersion(localMeta, dataKey);

  const remoteEntry = remoteStatuses.find((s) => s.dataKey === dataKey);
  const remoteVersion = remoteEntry?.dataVersion ?? 0;

  // No conflict if versions match
  if (localVersion === remoteVersion) {
    return null;
  }

  // No local data — just need to pull, not a conflict
  if (localVersion === 0) {
    return null;
  }

  // No remote data — safe to push
  if (remoteVersion === 0) {
    return null;
  }

  // Versions diverge — conflict
  return {
    localVersion,
    remoteVersion,
    dataKey,
  };
}

File: packages/cli/crates/server/src/sync/merge.rs
use swixter_core::types::SyncMeta;

use crate::sync::types::{SyncConflict, SyncStatusEntry};

fn local_version(meta: Option<&SyncMeta>, data_key: &str) -> u64 {
    match (meta, data_key) {
        (Some(m), "config") => m.config_version,
        (Some(m), "providers") => m.providers_version,
        _ => 0,
    }
}

/// TS: sync/merge.ts detectConflict
/// local==remote 或任一方为 0 → 无冲突；双方非零且不等 → 冲突。
pub fn detect_conflict(
    local_meta: Option<&SyncMeta>,
    remote_statuses: &[SyncStatusEntry],
    data_key: &str,
) -> Option<SyncConflict> {
    let local = local_version(local_meta, data_key);
    let remote = remote_statuses
        .iter()
        .find(|s| s.data_key == data_key)
        .map(|s| s.data_version)
        .unwrap_or(0);
    if local == remote || local == 0 || remote == 0 {
        return None;
    }
    Some(SyncConflict {
        local_version: local,
        remote_version: remote,
        data_key: data_key.to_string(),
    })
}

/// 远端某 dataKey 的当前版本（无则 0）——push 的 dataVersion 用它（乐观锁）
pub fn remote_version(statuses: &[SyncStatusEntry], data_key: &str) -> u64 {
    statuses
        .iter()
        .find(|s| s.data_key == data_key)
        .map(|s| s.data_version)
        .unwrap_or(0)
}
